core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java

```diff
@@ -0,0 +1,44 @@
+package io.quarkus.runtime.annotations;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation that can be used to force a class to be registered for serialization in native image mode
+ *
+ * Note that all class hierarchy will be registered. If you need a alternative for reflection only RegisterForReflection
+ * annotation
+ *
+ * @see RegisterForReflection
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface RegisterForSerialization {
+
+    /**
+     * If the methods should be registered
+     */
+    boolean methods() default true;
+
```

Comment:
The code is manipulating the methods in order to process the return and parameter types. I need to do some testing to be sure it's not needed. But what you mentioned makes sense

Reply:
@igorregis can you please clarify to which code you are referring to when saying "the code"?

Reply:
Sorry for that @zakkak, I was referring to [this method](https://github.com/quarkusio/quarkus/pull/24474/files#diff-152ea78045c4169902158c274ce4cd6e4d9a79f0676f71dfa1e6c1dc969300a8R128-R144). Since it's trying to deal only with getters and setters, it's seems that the "method" property could be removed from the annotation, but this method should be maintained so the change can behave the way you commented here:

> 2. Introduce RegisterForSerialization (as it already does) which should register the class and its hierarchy for reflection including all fields (ideally excluding transient ones) and ideally excluding all methods, except for writeObject, readObject and readObjectNoData, as they are not needed for serialization.

Reply:
I am still not convinced that serialization/deserialization requires reflective access to all potential getters and setters. The implementation should (at least in theory) only rely on the `writeObject`, `readObject` and `readObjectNoData` methods (see [here](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/Serializable.html)). Anything else should be explicitly registered by the user IMO. @Sanne WDYT?

Reply:
sorry I don't have experience with Serialization in native. Sounds like adding some tests would be nice?

BTW that's why I was suggesting to keep the builditems separate, one for each purpose: if - as a user of wuch build items - I'm needing *serialization* I shouldn't be using a builditem for *reflection* - if indeed serialization required reflection under the hood, that's an implementation detail.

Reply:
I may be missing something here, but as fas as I undertood it, those methods are optional ([see here](https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/serial-arch.html#the-serializable-interface)) and should be implemented only if the serializable object need something special for its serialization process.

Reply:
Hi @igorregis, you are right. Defining these methods is optional **but** when they are defines they need to be registered for reflection. The reflective access of these methods (when present) happens [here](https://github.com/openjdk/jdk17u-dev/blob/4b1bba808e659dc5ad18d1696bfe564a1df6d3e9/src/java.base/share/classes/java/io/ObjectStreamClass.java#L412-L421)

Reply:
I think we shouldn't mix Serialization with the original intent of the issue, which is, being able to marshall / unmarshall (usually by Jackson or other binding libraries that rely on reflection) a specific object and its components without having to know the entire object structure.

Reply:
Sorry for the delay, I've been on work trip that took all of my free time. Based on all of that discussion, this implementation should:
a) Enable a class to be correctly registered to be manipulated through reflection **OR**
b) Enable a class to be correctly registered to be serialized **OR**
c) Enable a class to be manipulated through reflection so its state can be retrieved and transferred through some IO process.

I understand that you are pointing on c) option @radcortez.

I have never read the source code from a library like Jackson deep enough to see if they comply with class level serialization or if they introspect the field values through reflection. I suspect it's the later option, based on tests results from CI. But I would need to do some research  in order to confirm that, if you think it's a good idea.

By any case, the problem may be the annotation name that need to be chosen in order to express the final intent.
Based on the three options above we have:
a) @RegisterForReflection
b) @RegisterForSerialization
c) @RegisterForObjectMapping **or** @RegisterForDataBiding **or even** @RegisterForJSONSerialization

Reply:
Yes, I stumbled upon this when doing the JSON marshaling / unmarshalling, but I don't want to make this a specific thing of JSON. We probably need to take one step back and look into the originally reported enhancement.

When registering a class for reflection using `-H:ReflectionConfigurationFiles` and GraalVM configuration style, the entire class hierarchy is added. When using our annotation, only the specified class is registered, and you need to explore all the classes to ensure everything is registered properly which is inconvenient.

The proposed enhancement was to make the register of reflection via the file vs via the annotation more consistent and alike.

I believe that the Serialization discussion is a completely different topic and should not be mixed with the original feature enhancement proposal.

Reply:
> I believe that the Serialization discussion is a completely different topic and should not be mixed with the original feature enhancement proposal.

I completely agree